### Problem 1 (10 Points)

Sketch graphically the problem 

$$
\begin{aligned}
\min_{x_1,x_2} & \quad f({\bf x})=(x_1+1)^2+(x_2-2)^2\\
{\text{subject to }} & \quad g_1 = x_1-2\leq 0,{\quad} g_3 = -x_1\leq 0,\\
& \quad g_2 = x_2-1\leq 0, {\quad} g_4 = -x_2\leq 0.
\end{aligned}
$$

Find the optimum graphically. Determine directions of feasible descent at the corner points of the feasible domain. Show the gradient directions of $f$ and $g_i$s at these points. Verify graphical results analytically using the KKT conditions.

---
### Solution
---

First we plot the ojective function and the constraints. For the objective function, we plot its contour lines. 

In [ ]:
import matplotlib.pyplot as plt
from matplotlib.pyplot import figure
import numpy as np


X = np.linspace(-1, 2, 1000)
Y = np.linspace(-1, 2, 1000)

X, Y = np.meshgrid(X, Y)

Z = (X+1)**2 + (Y-2)**2


fig, axs = plt.subplots(1, 1, figsize=(10, 6))
cp = plt.contour(X, Y, Z, levels= [i for i in range(0, 20, 2)])
#plt.clabel(cp, inline=1, fontsize=10)
plt.xlabel('X1')
plt.ylabel('X2')



y = np.linspace(-1,2.5)
plt.plot(2*np.ones(len(y)), y, c='red')
plt.plot(y, np.ones(len(y)), c = 'red')
plt.plot(np.zeros(len(y)), y, c='red')
plt.plot(y, np.zeros(len(y)), c='red')
plt.ylim([-1, 1.5])


plt.clabel(cp, inline=True, fontsize=8)
plt.title("Contour of the objective function w/constraints")  
plt.scatter(2, 0, marker='x', s=80, color='blue', label='maximum')
plt.scatter(0,1, marker='o', s=80, color='magenta', label='minimum')

x = np.linspace(0, 2, 100)
y = np.linspace(0, 1, 100)
plt.fill_between(x, y, 0, color='green', label='feasible-region')
plt.fill_between(x, y, 1, color='green')
plt.legend(loc='upper right')



plt.show()



From the plot above, we can see that the minimizer is $(x_1, x_2) = (0, 1)$ and the maximizer is $(x_1, x_2) = (2, 0)$

### Direction of feasible descent at the corner points: 

To determine the direction of feasible descents at the corner points we take the gradient of the ojective function first and evaluate at the corner points. Then we find the gradient of the activate constraints at the same points.


$$
\begin{aligned}
\quad\nabla f({\bf x})=\begin{bmatrix} 2x_1 + 2 \\
                                       2x_2 -4 \\
                                       \end{bmatrix}\\
\end{aligned}
$$

The gradients of the constraints are:

So, 
$$
\begin{aligned}
\quad\nabla g_1 &=\begin{bmatrix} 1 \\
                                 0 \\
                                       \end{bmatrix}\\  
\quad\nabla g_2 &=\begin{bmatrix} 0 \\
                                 1 \\
                                       \end{bmatrix}\\                      
\quad\nabla g_3 &=\begin{bmatrix} -1 \\
                                 0 \\
                                       \end{bmatrix}\\
\quad\nabla g_4 &=\begin{bmatrix} 0 \\
                                 -1 \\
                                       \end{bmatrix}\\  
\end{aligned}
$$


The active constraints at the minimizer $(0, 1)$ are: $g_2\;\text{and}\;g_3$ and the gradient of the objective function is:

$$
\begin{aligned}
\quad\nabla f({\bf x})=\begin{bmatrix} 2 \\
                                       -2 \\
                                       \end{bmatrix}\\
\end{aligned}
$$

Similarly, the active constraints at the maximizer $(2, 0)$ are: $g_1\;\text{and}\;g_4$ and the gradient of the objective function is:
$$
\begin{aligned}
\quad\nabla f({\bf x})=\begin{bmatrix} 6 \\
                                       -4 \\
                                       \end{bmatrix}\\
\end{aligned}
$$

Plotting the above gradients in the above plot:

In [ ]:
import matplotlib.pyplot as plt
from matplotlib.pyplot import figure
import numpy as np


X = np.linspace(-1, 2, 1000)
Y = np.linspace(-1, 2, 1000)

X, Y = np.meshgrid(X, Y)

Z = (X+1)**2 + (Y-2)**2


fig, axs = plt.subplots(1, 1, figsize=(10, 6))
cp = plt.contour(X, Y, Z, levels= [i for i in range(0, 20, 2)])f
#plt.clabel(cp, inline=1, fontsize=10)
plt.xlabel('X1')
plt.ylabel('X2')



y = np.linspace(-1,2.5)
plt.plot(2*np.ones(len(y)), y, c='brown', linestyle='dashed')
plt.plot(y, np.ones(len(y)), c='brown', linestyle='dashed')
plt.plot(np.zeros(len(y)), y, c='brown', linestyle='dashed')
plt.plot(y, np.zeros(len(y)), c='brown', linestyle='dashed')
plt.ylim([-1, 1.5])


plt.clabel(cp, inline=True, fontsize=8)
plt.title("Gradient direction of the objective function and contraints at the optimum points")  
plt.scatter(2, 0, marker='x', s=200, color='blue', label='maximum')
plt.scatter(0,1, marker='o', s=200, color='magenta', label='minimum')

x = np.linspace(0, 2, 100)
y = np.linspace(0, 1, 100)
plt.fill_between(x, y, 0, color='green', label='feasible-region')
plt.fill_between(x, y, 1, color='green')
plt.legend(loc='upper right')

origin1 = np.array([0, 1])
plt.quiver(*origin1, [1, 0, -1], [-1, 1 ,0], color=['black', 'red', 'green'], scale=10)

origin2 = np.array([2, 0])
plt.quiver(*origin2, [1.5, 1, 0], [-1, 0 ,-1], color=['black', 'red', 'green'], scale=10)

origin3 = np.array([0, 0])
plt.quiver(*origin3, [1, 1, 0], [-2, 0 ,-1], color=['black', 'red', 'green'], scale=10)

origin4 = np.array([2, 1])
plt.quiver(*origin4, [1.5, 1, 0], [-0.5, 0 ,-1], color=['black', 'green', 'red'], scale=10)

plt.show()

---------------------------------------------------------

**The optimum value is $f = 2$ at $(x_1, x_2) = (0, 1)$**

----------------------------------------------------------

### Analytical Solution Using KKT conditions:
---

For KKT conditions, we need Lagrangian, which is defined as:

$$
\begin{aligned}
\quad \mathcal{L} =\quad f + \mu_1 g_1 + \mu_2 g_2 + \mu_3 g_3 + \mu_4 g_4 \\
\end{aligned}
$$

where, $\mu_i$s are the dual variables (aka lagrange multipliers) for the inequality constraints $g_i$s.

$$
\begin{aligned}
\quad \mathcal{L} = (x_1+1)^2+(x_2-2)^2 + \mu_1 (x_1-2) + \mu_2 (x_2-1) + \mu_3 (-x_1) + \mu_4 (-x_4) \\
\end{aligned}
$$

Then, the gradient of the Lagrangian is:

$$
\begin{aligned}
\quad \nabla \mathcal{L} = \begin{bmatrix} 2x_1 + 2 + \mu_1 - \mu_3\\
                                       2x_2 -4 + \mu_2 - \mu_4 \\
                                       \end{bmatrix}\\
\end{aligned}
$$

Set $\nabla \mathcal{L} = 0$

$$
\begin{aligned}
\begin{bmatrix} 2x_1 + 2 + \mu_1 - \mu_3\\
                                       2x_2 -4 + \mu_2 - \mu_4 \\
                                       \end{bmatrix} &= \begin{bmatrix} 0 \\
                                                                        0 \end{bmatrix}\\
\end{aligned}
$$


Then, at the minimum, taking $g_2$ and $g_3$ as activate constraints (usually we can guess when we don't have any representation of the objective function). Then, using the matrix above along with the active constraints, we can solve for $\mu_i$s (note: $\mu_1 = \mu_4 = 0$):

We get $\mu_2 = 2$, $\mu_3 = 2$, $x_1 = 0$ and $x_2 = 1$. 

As $\mu_2 > 0$ and $\mu_3 > 0$, and the feasible region is convex, <b> the minimum occurs at (0, 1) and the minimum value is 2. </b>

#### Hence, the result is consistent with the graphical method.

---
---


### Problem 2 (10 Points)

Graph the problem 

$$
\begin{aligned}
\min_{x_1,x_2} & \quad  f=-x_1\\
{\text{subject to }} & \quad g_1=x_2-(1-x_1)^3\leq 0{\quad} {\rm and}{\quad} x_2\geq 0.
\end{aligned}
$$ 

Find the solution graphically. Then apply the optimality conditions. Can you find a solution based on the optimality conditions? Why? (From Kuhn and Tucker, 1951.)

---
Solution: Graphically
---

In [ ]:
x = np.linspace(-1, 1.4, 100)
y = np.linspace(-1, 2, 100)
g = (1-x)**3


X = np.linspace(-1, 1.4, 100)
Y = np.linspace(-1,8,100)
X, Y = np.meshgrid(X, Y)

Z = -X


fig, axs = plt.subplots(1, 1, figsize=(10, 6))
cp = plt.contour(X, Y, Z)

plt.plot(x, g, label = 'g1')
plt.plot(x, np.zeros(len(x)), linestyle='dashed')

plt.fill_between(x, g, 0, color='yellow', label='feasible-region')
plt.xlabel("x1")
plt.ylabel("x2")
plt.legend()
plt.title("Objective function w/constraints and feasible region")

From the plot above, we can observe that the minimum value for the given constraints occurs at $\mathbf{(x_1, x_2) = (1, 0)}$. 

### Solution using optimality condition

The Lagrangian of the optimization problem is: 

$$
\begin{aligned}
\quad \mathcal{L} &=\quad f + \mu_1 g_1 + \mu_2 (-x_2) \\
                  &=\quad -x_1 + \mu_1 (x_2-(1-x_1)^3)  + \mu_2 (-x_2) \\
\end{aligned}
$$

Then, the gradient of the Lagrangian is:

$$
\begin{aligned}
\quad \nabla \mathcal{L} = \begin{bmatrix} 3\mu_1(1 - x_1)^2 -1\\
                                       \mu_1 - \mu_2 \\
                                       \end{bmatrix}\\
\end{aligned}
$$

Set $\nabla \mathcal{L} = 0$

$$
\begin{aligned}
\begin{bmatrix} 3\mu_1(1 - x_1)^2 -1\\
                                       \mu_1 - \mu_2 \\
                                       \end{bmatrix} &= \begin{bmatrix} 0 \\
                                                                        0 \end{bmatrix}\\
\end{aligned}
$$

Set $g_2$ as active constraint, then $\mu_1 = 0, \mu_2 > 0$. However, $\mu_2 = 0$ from the above relationship. 


We can determine the gradient of the objective function and constraints at (1,0). 

$$
\begin{aligned}
\quad \nabla f_{(1, 0)} = \begin{bmatrix} -1 \\ 0 \end{bmatrix}\\
\quad \nabla g_{1 (1, 0)} = \begin{bmatrix} 0 \\ 1 \end{bmatrix}\\
\quad \nabla g_{2 (1,0)} = \begin{bmatrix} 0 \\ -1 \end{bmatrix}\\
\end{aligned}
$$

We can see that the gradient of the constraints are linearly dependent, and hence we cannot determine the optimal solution using the optimality condition. We can plot these vectors to see their behavior. 

In [ ]:
x = np.linspace(-1, 1.4, 100)
y = np.linspace(0, 2, 100)
g = (1-x)**3


X = np.linspace(-1, 1.4, 100)
Y = np.linspace(-1,8,100)
X, Y = np.meshgrid(X, Y)

Z = -X


fig, axs = plt.subplots(1, 1, figsize=(10, 6))
cp = plt.contour(X, Y, Z)

plt.plot(x, g, label = 'g1')
plt.plot(x, np.zeros(len(x)), linestyle='dashed')

plt.fill_between(x, g, 0, color='yellow', label='feasible-region')
plt.xlabel("x1")
plt.ylabel("x2")
plt.legend()
plt.title("Objective function w/constraints and feasible region")
origin1 = np.array([1, 0])
plt.quiver(*origin1, [-1, 0, 0], [0, 1 ,-1], color=['black', 'red', 'green'], scale=15)

### Problem 3 (30 Points)

Find a local solution to the problem 

$$
\begin{aligned}
\max_{x_1,x_2,x_3} & \quad  f=x_1x_2+x_2x_3+x_1x_3\\
{\text{subject to }} & \quad h=x_1+x_2+x_3-3=0.
\end{aligned}
$$ 

Use two methods: reduced gradient and Lagrange multipliers.

### Solution 1: Using Reduced Gradient
---

In reduced gradient method, first, we need to identify the number of constraints (m). Here, m = 1.

So, there must be n - m decision variables, where n = 3. Hence, $d \in \mathbb{R}^2$ and $s \in \mathbb{R}^1$

We pick $x_1$ as state variable and $x_2, x_3$ as decision variables. Then: 

$x = \begin{bmatrix}d \\ s\end{bmatrix}$

Now, 

$$
\begin{aligned}
\frac{\partial f}{\partial d} &= \begin{bmatrix} -x_1 - x_3 \\ -x_1 - x_2\end{bmatrix}\\
\frac{\partial f}{\partial s} &= -x_2 - x_3 \\
\frac{\partial h}{\partial d} &= \begin{bmatrix} 1 \\ 1\end{bmatrix}\\
\frac{\partial h}{\partial s} &= 1\\ \\
\text{And,}\\
\frac{df}{dd} &= \frac{\partial f}{\partial d} - \left(\frac{\partial f}{\partial s}\right)\left(\frac{\partial h}{\partial s} \right)^{-1} \frac{\partial h}{\partial d}\\
              &= \begin{bmatrix} -x_1 - x_3 \\ -x_1 - x_2\end{bmatrix} - (-x_2 - x_3) (1) \begin{bmatrix} 1 \\ 1\end{bmatrix}\\
              &= \begin{bmatrix} -x_1 - x_3 \\ -x_1 - x_2\end{bmatrix} - \begin{bmatrix} -x_2 - x_3 \\ -x_2 - x_3\end{bmatrix}\\
              &= \begin{bmatrix} x_2 - x_1 \\ x_3 - x_1\end{bmatrix}\\
\end{aligned}
$$

Set $\frac{df}{dd} = 0$. We get: $x_1 = x_2 = x_3$. Plugging these back into the constraints results in:

$$
\begin{aligned}
x_1^* = x_2^* = x_3^* = 1 \;\;\text{and}\;\; f^* = 3
\end{aligned}
$$



<!-- To verify that the above solutions are indeed local minimum we need to check the second order sufficient condition.  


We know, 

$$
\begin{aligned}
\quad \frac{{\rm d}s}{{\rm d}d} &= -\left(\frac{\partial h}{\partial s}\right)^{-1} \frac{\partial h}{\partial d} \\ 
& = \begin{bmatrix} -1 \\ -1 \end{bmatrix}\\
\end{aligned}
$$

Then, We can see, $\frac{{\rm d^2} s}{{\rm d} d^2} = 0$.

Now,

$$
\begin{aligned}
\quad \frac{{\rm d}^2 f}{{\rm d}d^2} &= 
\begin{bmatrix}
I & \left(\frac{{\rm d}s}{{\rm d}d}\right)^T\\
\end{bmatrix}
\begin{bmatrix}
\frac{{\partial}^2 f}{{\partial}d^2} & \frac{{\partial}f}{{\partial}d{\partial}s}\\
\frac{{\partial}f}{{\partial}s{\partial}d} & \frac{{\partial}^2 f}{{\partial}s^2}\\
\end{bmatrix}
\begin{bmatrix}
I \\
\frac{{\rm d}s}{{\rm d}d}\\
\end{bmatrix}
+
\frac{{\partial}f}{{\partial}s}\frac{{\rm d}^2 s}{{\partial}d^2} \\ 
& =
\begin{bmatrix}
1&-1&-1 \\
\end{bmatrix}
\begin{bmatrix}
0&1&1 \\
1&0&1 \\
1&1&0 \\
\end{bmatrix}
\begin{bmatrix}
1 \\
-1 \\
-1 \\
\end{bmatrix} = -2 < 0\\
\end{aligned}
$$

Its not the minimizer.  -->


<!-- $$
\begin{aligned}
\quad & \frac{\rm d^2 h}{\rm d d^2} = 
\begin{bmatrix}
I&\left(\frac{{\rm d}s}{{\rm d}d}\right)^T\\
\end{bmatrix}
\begin{bmatrix}
\frac{{\partial}^2 h}{{\partial}d^2} & \frac{{\partial}^2h}{{\partial}d{\partial}s}\\
\frac{{\partial}^2h}{{\partial}s{\partial}d} & \frac{{\partial}^2 h}{{\partial}s^2}\\
\end{bmatrix}
\begin{bmatrix}
I \\
\frac{{\rm d}s}{{\rm d}d}\\
\end{bmatrix}
+
\frac{{\partial}h}{{\partial}s}\frac{{\rm d}^2 s}{{\rm d}d^2} = 0\\
\end{aligned}
$$

 -->

<!-- Then, 
$$
\begin{aligned}
\quad & \frac{{\partial}^2 f}{{\partial}d^2} = 
\begin{bmatrix}
0 & 1 \\
1 & 0 \\
\end{bmatrix}\\
\quad & \frac{{\partial}^2 f}{{\partial}s^2} = 0\\
\quad & \frac{{\partial}f}{{\partial}s{\partial}d} = 
\begin{bmatrix}
1 & 1 \\
\end{bmatrix}\\
\quad & \frac{{\partial}f}{{\partial}d{\partial}s} = 
\begin{bmatrix}
1 \\
1 \\
\end{bmatrix}\\
\end{aligned}
$$ -->

---
---
### Solution 2: Using Lagrange Multipliers


The Lagrangian for the optimization problem is $\mathcal{L} = -f + \lambda_1 h$, where $h$ is the equality constraint. 

$$
\begin{aligned}
\quad \mathcal{L} &= -x_1x_2 - x_2x_3 - x_1x_3 + \lambda_1 (x_1 + x_2 + x_3 - 3)\\
\text{Then, gradient of the Lagrangian is:}\\
\quad \nabla \mathcal{L} &= \begin{bmatrix} -x_2 - x_3 + \lambda_1 \\
                                            -x_1 - x_2 + \lambda_1 \\
                                            -x_1 - x_3 + \lambda_1
\end{bmatrix} = 0
\end{aligned}
$$

On solving, we get:

$x_1 = x_2 = x_3 = 1, \lambda = 2$.

To further verify the optimality, we need to check for second order sufficiency condition.

<!-- $$
\begin{aligned}
\quad \frac{{\rm d}^2\mathcal{L}}{{\rm d}x^2} &= \begin{bmatrix} 0 & 1 & 1 & 1\\
                                                                 1 & 0 & 1 & 1\\
                                                                 1 & 1 & 0 & 1\\
                                                                 1 & 1 & 1 & 0\\\end{bmatrix}\;\;\text{p.d}
\end{aligned}
$$ -->


$$
\begin{aligned}
\partial x^T\mathcal{L_{xx}}\partial x &= \begin{bmatrix}dx_1 & dx_2 & dx_3 \end{bmatrix} \begin{bmatrix} 0 & 1 & 1\\
                                                                 1 & 0 & 1\\
                                                                 1 & 1 & 0\\
                                                                 \end{bmatrix} \begin{bmatrix}dx_1 \\ dx_2 \\ dx_3 
                                                                 \end{bmatrix}
                                        &= -(dx_1^2 +dx_2^2 - dx_1dx_2)\\
                                        &= -\left(\left(dx_1 - \frac{dx_2}{2}\right)^2 + \frac{4}{3}dx_2^2\right) \le 0
\end{aligned}
$$

Hence, $x_1^* = x_2^* = x_3^* = 1$ is the global maximizer. 

---
---

### Problem 4 (20 Points)

Use reduced gradient to	find the value(s) of the parameter $b$ for which the point $x_1=1$, $x_2=2$ is the solution to the problem 

$$
\begin{aligned}
\max_{x_1,x_2} & \quad  f=2x_{1} + bx_2\\
{\text{subject to }} & \quad g_1 = x_{1}^{2}+ x_{2}^{2}-5\leq 0\\
& \quad g_2= x_1- x_2-2\leq 0.
\end{aligned}
$$ 

### Solution: 

At $x_1 = 1$ and $x_2 = 2$, set $g_1$ as the active constraint. Then, assume $x_1$ is the state variable and $x_2$ is the decision variable. Then, 

$$
\begin{aligned}
\quad \frac{\partial f}{\partial d} &= b \\
\quad \frac{\partial f}{\partial s} &= 2 \\
\quad \frac{\partial h}{\partial d} &=  \frac{\partial g_1}{\partial d} = 2x_2 \\
\quad \frac{\partial h}{\partial s} &= \frac{\partial g_1}{\partial s} = 2x_1 \\
\end{aligned}
$$

Then, 

$$
\begin{aligned}
\frac{df}{dd} &= \frac{\partial f}{\partial d} - \left(\frac{\partial f}{\partial s}\right)\left(\frac{\partial h}{\partial s} \right)^{-1} \frac{\partial h}{\partial d}\\
              &= b - 2 \left(\frac{1}{2x_1}\right)2x_2 \\
              &= b - \frac{2x_2}{x_1}
\end{aligned}
$$
Then, at $x_1 = 1$ and $x_2 = 2$, 

$$
\begin{aligned}
\frac{df}{dd} &= b - \frac{2\times2}{1} = 0 \\
\text{Hence,}\\
              b &= 4\\
\end{aligned}
$$

We can check the second order sufficient condition as well. 

$$
\begin{aligned}
\quad \frac{{\rm d}^2 f}{{\rm d}d^2} &= 
\begin{bmatrix}
I & \left(\frac{{\rm d}s}{{\rm d}d}\right)^T\\
\end{bmatrix}
\begin{bmatrix}
\frac{{\partial}^2 f}{{\partial}d^2} & \frac{{\partial}f}{{\partial}d{\partial}s}\\
\frac{{\partial}f}{{\partial}s{\partial}d} & \frac{{\partial}^2 f}{{\partial}s^2}\\
\end{bmatrix}
\begin{bmatrix}
I \\
\frac{{\rm d}s}{{\rm d}d}\\
\end{bmatrix}
+
\frac{{\partial}f}{{\partial}s}\frac{{\rm d}^2 s}{{\rm d}d^2} \\ 
\end{aligned}
$$

where, 

$$
\begin{aligned}
\quad \frac{{\rm d}s}{{\rm d}d} &= -\left(\frac{\partial h}{\partial s}\right)^{-1} \frac{\partial h}{\partial d} \\ 
& = -\frac{1}{2x_1} 2x_2\\
& = -\frac{x_2}{x_1}\\
\end{aligned}
$$

$$
\begin{aligned}
\quad \frac{{\rm d}^2 h}{{\rm d}d^2} &= 
\begin{bmatrix}
I & \left(\frac{{\rm d}s}{{\rm d}d}\right)^T\\
\end{bmatrix}
\begin{bmatrix}
\frac{{\partial}^2 h}{{\partial}d^2} & \frac{{\partial}h}{{\partial}d{\partial}s}\\
\frac{{\partial}h}{{\partial}s{\partial}d} & \frac{{\partial}^2 h}{{\partial}s^2}\\
\end{bmatrix}
\begin{bmatrix}
I \\
\frac{{\rm d}s}{{\rm d}d}\\
\end{bmatrix}
+
\frac{{\partial}h}{{\partial}s}\frac{{\rm d}^2 s}{{\rm d}d^2} = 0 \\ 
\end{aligned}
$$

$$
\begin{aligned}
&= 
\begin{bmatrix}
1 & -2\\
\end{bmatrix}
\begin{bmatrix}
2 & 0\\
0 & 2\\
\end{bmatrix}
\begin{bmatrix}
1 \\
-2\\
\end{bmatrix}
+
2\frac{{\rm d}^2 s}{{\rm d}d^2} = 0 \\ 
\end{aligned}
$$

$$
\begin{aligned}
\frac{{\rm d}^2 s}{{\rm d}d^2} &=  -5
\end{aligned}
$$

So, 

$$
\begin{aligned}
\quad \frac{{\rm d}^2 f}{{\rm d}d^2} &= \frac{{\partial}f}{{\partial}s}\frac{{\rm d}^2 s}{{\partial}d^2}\;\;\;\left(\text{Since}\;\;f_{xx} = 0 \right)
&= 2 (-5) = -10 
\end{aligned}
$$

Therefore, b=4 maximizes the objective and the maximum value is: 2 

---



### Problem 5 (30 Points)

Find the solution for 

$$
\begin{aligned}
\min_{x_1,x_2,x_3} & \quad  f=x_{1}^{2}+x_{2}^{2}+x_{3}^{2}\\
{\text{subject to }} & \quad h_1 = x_{1}^{2}/4+x_{2}^{2}/5+x_{3}^{2}/25-1=0\\
& \quad h_2 = x_1+x_2-x_3= 0,
\end{aligned}
$$ 

by implementing the generalized reduced gradient algorithm.

---
---

### Solution

For generalized reduced gradient alogirthm, we need to separate the space into state and decision spaces. Since there are 2 constraints, we need to state variables and one decision variables. We pick $x_1$ as decision variable and $x_2$ and $x_3$ as state variables. Following program is used to calculate the reduced gradient using the GRG algorithm paired with the line search algorithm to find the optimum step size at each iteration. 

In [12]:
import torch as t
import numpy as np

# objective function
f = lambda x: x[0][0]**2 + x[1][0]**2 + x[2][0]**2

# constraints 
def f_g(x):
    return np.array([[2*x[0][0]],[2*x[1][0]],[2*x[2][0]]]);


# m = 2, n = 3, n-m = 1; 

def const(x):
    return np.array([[(x[0][0]**2)/4 + (x[1][0]**2)/5 + (x[2][0]**2)/25 - 1], [x[0][0] + x[1][0] - x[2][0]]])

def constg(x):
    return np.array([[x[0][0]/2, 2*x[1][0]/5, 2*x[2][0]/25],[1, 1, -1]])


# decision var -- x1, x2 & x3 are state
def solveh(x, const, constg):
    import matplotlib.pyplot as plt
    error = 1e-2
    
    s = x[1:] # state 
    
    #check if the constraint is satisfied
    
    h = const(x)
     # take pseudo inverse
    norm = []
    count = 0
    while np.linalg.norm(h) >= error and count < 20:
        count += 1
        dh_dx =  constg(x)
        dh_ds = dh_dx[:,1:]
        dh_inv = np.linalg.pinv(dh_ds)
        s = s - np.matmul(dh_inv, h)
        x[1] = s[0][0]
        x[2] = s[1][0]
        h = const(x)
        norm.append(np.linalg.norm(h))
    
    #plt.plot(norm)
    return x
    

def df_dd(x):
    par_fd = f_g(x)[0].reshape(1,1)
    par_fs = f_g(x)[1:]
    dh_dx =  constg(x)
    dh_ds = dh_dx[:,1:]
    dh_dd = dh_dx[:, 0].reshape(2,1)
    dh_inv = np.linalg.inv(dh_ds)

    
    #df_dd = f_g(x)[0] - np.matmul(np.matmul(f_g(x)[1:, :].T, dh_inv), dh_dx[:,0].reshape(2,1))
    
    df_dd1 = par_fd - np.matmul(par_fs.T, np.matmul(dh_inv, dh_dd))
    
    return df_dd1, dh_inv, dh_ds, dh_dd
    
def line_search(x):
    a = 1. 
    tilt = 0.3
    dfdd, _, _, _ = df_dd(x)
    phi = lambda x, a, tilt, dfdd: f(x) - a * tilt * np.matmul(dfdd, dfdd.T)

    def f_a(x, alpha, tilt):
        dfdd, dh_inv, dh_ds, dh_dd = df_dd(x)
        
        x[0][0] = x[0][0] - alpha*dfdd
        x[1:] = x[1:] + alpha*np.matmul(np.matmul(np.linalg.inv(dh_ds), dh_dd), dfdd.T)
        return f(x) 
                                     
                                      
    while phi(x, a, tilt, dfdd) < f_a(x, a, tilt):
        a *= 0.5
        dfdd, _, _, _ = df_dd(x)
    return a   

    
def GRG(x, const, constg):
    
    error = 1e-3
    
    df_dd1, dh_inv, dh_ds,_ = df_dd(x)
    h = const(x)
    count = 0
    norms = []
    I = np.diag(np.ones((dh_ds.shape[0])))
    lam = 1
    xs = []
    while (np.linalg.norm(df_dd1) >= error):
        x1 = x.copy() # don't want to change the initial x --> pass by reference
        alpha = line_search(x1)
        norms.append(np.linalg.norm(df_dd1))
        count += 1
        x[0][0] -= alpha * df_dd1
        s = x[1:]

        s = s - np.matmul(np.linalg.inv(np.matmul(dh_ds.T, dh_ds)+lam*I), np.matmul(dh_ds.T, h))
                
        x[1:] = s
        x = solveh(x, const, constg)
        
        df_dd1, dh_inv, dh_ds,_ = df_dd(x)
        h = const(x)
        xs.append(x)
        
    return x, norms, xs
        

In [21]:
# initial guess
x = np.array([[1.,2.,3.]]).T

x_min, norms, xs = GRG(x, const, constg)

In [22]:
x_min

array([[-1.5753846 ],
       [ 1.37773504],
       [-0.19693537]])

<b>Hence, the minimizer for the given problem is: $\left(x_1^*, x_2^*, x_3^*\right) = \left(-1.5754, 1.37773, -0.19694\right)$</b> 